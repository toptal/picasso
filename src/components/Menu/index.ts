export { default, MenuContext } from './Menu'
