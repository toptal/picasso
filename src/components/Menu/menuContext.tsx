import { createContext } from 'react'

import { MenuContextProps } from './types'

export default createContext<MenuContextProps>({} as MenuContextProps)
