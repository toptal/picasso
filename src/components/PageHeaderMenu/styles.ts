import { createStyles } from '@material-ui/core/styles'

export default () =>
  createStyles({
    avatar: {
      fontSize: '0.9rem'
    },
    content: {
      width: '15em'
    },
    name: {
      fontWeight: 400,
      display: 'block'
    },
    truncateText: {
      maxWidth: '11.5rem',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis'
    }
  })
