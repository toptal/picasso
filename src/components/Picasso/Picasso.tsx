import {
  createMuiTheme,
  MuiThemeProvider,
  withStyles
} from '@material-ui/core/styles'
import React, { FunctionComponent, ReactNode } from 'react'

import CssBaseline from '../CssBaseline'
import {
  palette,
  layout,
  transitions,
  typography,
  spacing,
  breakpoints,
  screens,
  shadows
} from './config'
import FontsLoader from './FontsLoader'
import Provider from './PicassoProvider'
import globalStyles from './styles'
import { JssProps } from './types'

const picasso = {
  palette,
  layout,
  transitions,
  spacing,
  breakpoints,
  screens,
  shadows,
  typography,
  props: {
    MuiButtonBase: {
      disableRipple: true
    },
    MuiList: {
      disablePadding: true
    },
    MuiPaper: {
      square: true
    },
    MuiOutlinedInput: {
      notched: false
    }
  }
}

const PicassoProvider = new Provider(createMuiTheme(picasso))

interface PicassoGlobalStylesProviderProps extends JssProps {
  children?: ReactNode
}

const PicassoGlobalStylesProvider = withStyles(globalStyles, {
  name: 'Picasso'
})((props: PicassoGlobalStylesProviderProps) => {
  const { classes, children } = props

  return <div className={classes.root}>{children}</div>
})

interface PicassoProps {
  children?: ReactNode
  /** Whether to load fonts file to the page */
  loadFonts?: boolean
  /** Whether to apply Picasso CSS reset */
  reset?: boolean
}

const Picasso: FunctionComponent<PicassoProps> = ({
  loadFonts,
  reset,
  children
}) => (
  <MuiThemeProvider theme={PicassoProvider.theme}>
    {loadFonts && <FontsLoader />}
    {reset && <CssBaseline />}
    <PicassoGlobalStylesProvider>{children}</PicassoGlobalStylesProvider>
  </MuiThemeProvider>
)

Picasso.defaultProps = {
  loadFonts: true,
  reset: true
}

export { PicassoProvider }
export default Picasso
