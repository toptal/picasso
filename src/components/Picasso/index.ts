export {
  default,
  PicassoProvider,
  usePicassoRoot,
  usePageHeader
} from './Picasso'

export {
  BaseProps,
  JssProps,
  StandardProps,
  SizeType,
  OmitInternalProps,
  SpacingType,
  spacingToEm,
  PicassoComponent,
  PicassoComponentWithRef,
  SpacingEnum,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  ColorType
} from './types'

export { useScreenSize, isScreenSize, useBreakpoint } from './config'
