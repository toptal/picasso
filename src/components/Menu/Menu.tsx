import React, {
  HTMLAttributes,
  forwardRef,
  useState,
  useContext,
  ReactElement
} from 'react'
import MUIMenuList, { MenuListProps } from '@material-ui/core/MenuList'
import { withStyles } from '@material-ui/core/styles'
import cx from 'classnames'

import { BackMinor16 } from '../Icon'
import {
  StandardProps,
  PicassoComponentWithRef,
  CompoundedComponentWithRef
} from '../Picasso'
import styles from './styles'
import Typography from '../Typography'
import MenuItem from '../MenuItem'
import MenuContext, { MenuContextProps } from './menuContext'

export type ListNativeProps = HTMLAttributes<HTMLUListElement> &
  Pick<MenuListProps, 'onKeyDown'>

export interface Props extends StandardProps, ListNativeProps {
  // whether or not to handle nested navigation
  allowNestedNavigation?: boolean
}

export interface StaticProps {
  Item: typeof MenuItem
}

type Menus = Record<string, ReactElement>

// eslint-disable-next-line react/display-name
export const Menu = forwardRef<HTMLUListElement, Props>(function Menu(
  { children, className, classes, style, allowNestedNavigation, ...rest },
  ref
) {
  const { backButtonIcon, hideMenu, ...restClasses } = classes
  const { pop } = useContext<MenuContextProps>(MenuContext)

  const hasParentMenu = !!pop

  const handleBackClick = (
    event: React.MouseEvent<HTMLElement, MouseEvent>
  ) => {
    event.stopPropagation()
    pop()
  }

  const menu = (
    <MUIMenuList
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={ref}
      className={className}
      style={style}
      classes={restClasses}
      // eslint-disable-next-line react/jsx-props-no-spreading
    >
      {hasParentMenu && allowNestedNavigation && (
        <MenuItem onClick={handleBackClick} key='back'>
          <Typography size='small' color='dark-grey' variant='body'>
            <BackMinor16 className={backButtonIcon} />
            Back
          </Typography>
        </MenuItem>
      )}
      {children}
    </MUIMenuList>
  )

  const [menus, setMenus] = useState<Menus>({})
  const menusKeys = Object.keys(menus)

  const getLastKey = () => menusKeys[menusKeys.length - 1]

  if (hasParentMenu) {
    return menu
  }

  const menuContext = {
    push: (key: string, menu: ReactElement) =>
      setMenus({ ...menus, ...{ [key]: menu } }),
    pop: () => {
      const key = getLastKey()

      if (!key) {
        return
      }

      const newMenus = { ...menus }

      delete newMenus[key]
      setMenus(newMenus)
    },
    refresh: (key: string, menu: ReactElement) => {
      if (!menus[key]) {
        return
      }

      setMenus({ ...menus, ...{ [key]: menu } })
    }
  } as MenuContextProps

  const currentVisibleMenuKey = getLastKey()
  const isRootMenuHidden = Boolean(currentVisibleMenuKey)

  return (
    <MenuContext.Provider value={menuContext}>
      {React.cloneElement(menu, {
        className: cx(menu.props.className, { [hideMenu]: isRootMenuHidden })
      })}
      {menusKeys.map((menuKey: string) =>
        React.cloneElement(menus[menuKey], {
          key: menuKey,
          className: cx(menus[menuKey].props.className, {
            [hideMenu]: menuKey !== currentVisibleMenuKey
          })
        })
      )}
    </MenuContext.Provider>
  )
}) as CompoundedComponentWithRef<Props, HTMLUListElement, StaticProps>

Menu.defaultProps = {
  allowNestedNavigation: true
}

Menu.displayName = 'Menu'

Menu.Item = MenuItem

export default withStyles(styles)(Menu) as PicassoComponentWithRef<
  Props,
  HTMLUListElement,
  StaticProps
>
