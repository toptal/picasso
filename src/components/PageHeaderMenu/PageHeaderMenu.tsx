import React, { FunctionComponent, ReactNode } from 'react'
import cx from 'classnames'
import { withStyles } from '@material-ui/core/styles'

import { StandardProps } from '../Picasso'
import UserBadge from '../UserBadge'
import Dropdown from '../Dropdown'
import Typography from '../Typography'
import styles from './styles'

export interface Props extends StandardProps {
  /** User full name to display */
  name: string
  /** User's organization name */
  organization?: string
  /** Photo url or custom Avatar component */
  avatar?: ReactNode
  /** Menu content */
  children: ReactNode
}

export const PageHeaderMenu: FunctionComponent<Props> = ({
  name,
  organization,
  avatar,
  classes,
  className,
  style,
  children
}) => {
  return (
    <Dropdown
      className={cx(classes.root, className)}
      classes={{ content: classes.content }}
      style={style}
      content={children}
      offset={{ top: 'xsmall' }}
    >
      <UserBadge
        invert
        center
        size='xsmall'
        classes={{
          avatar: classes.avatar,
          name: cx(classes.name, classes.truncateText)
        }}
        name={name}
        avatar={avatar}
      >
        {organization && (
          <Typography className={classes.truncateText} invert size='small'>
            {organization}
          </Typography>
        )}
      </UserBadge>
      <Dropdown.Arrow style={{ color: 'white' }} />
    </Dropdown>
  )
}

PageHeaderMenu.defaultProps = {}

PageHeaderMenu.displayName = 'PageHeaderMenu'

export default withStyles(styles)(PageHeaderMenu)
