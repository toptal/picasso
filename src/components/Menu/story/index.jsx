import PicassoBook from '~/.storybook/components/PicassoBook'

import menuItemStory from '@components/MenuItem/story'

import { Menu } from '../Menu'

const page = PicassoBook.createPage('Menu', `Menu list.`)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: Menu, name: 'Menu' })
  .addComponentDocs(menuItemStory.componentDocs)

page.createChapter().addExample('Menu/story/Default.example.jsx', 'Default')

page.connect(menuItemStory.chapter)
