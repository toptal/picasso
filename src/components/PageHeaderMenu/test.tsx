import React, { FunctionComponent } from 'react'
/* eslint-disable-next-line */
import { render, cleanup } from '@testing-library/react'

import Picasso, { OmitInternalProps } from '../Picasso'
import Menu from '../Menu'
import PageHeaderMenu, { Props } from './PageHeaderMenu'

const TestPageHeaderMenu: FunctionComponent<OmitInternalProps<Props>> = ({
  children,
  name,
  avatar
}) => (
  <Picasso loadFonts={false}>
    <PageHeaderMenu name={name} avatar={avatar}>
      {children}
    </PageHeaderMenu>
  </Picasso>
)

afterEach(cleanup)

describe('PageHeaderMenu', () => {
  test('default render', () => {
    const { container } = render(
      <TestPageHeaderMenu
        name='Jacqueline Roque'
        avatar='./jacqueline-with-flowers-1954-square.jpg'
      >
        <Menu>
          <Menu.Item>My Account</Menu.Item>
          <Menu.Item>Log Out</Menu.Item>
        </Menu>
      </TestPageHeaderMenu>
    )

    expect(container).toMatchSnapshot()
  })
})
