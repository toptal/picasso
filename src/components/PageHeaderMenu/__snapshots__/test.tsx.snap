// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeaderMenu default render 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="Dropdown-root"
    >
      <div
        class="Dropdown-anchor"
      >
        <div
          class="Container-flex UserBadge-root"
          style="align-items: center;"
        >
          <div
            class="Avatar-root Avatar-xsmall"
          >
            <img
              alt="Jacqueline Roque"
              class="Image-root Avatar-image Avatar-square Avatar-xsmall Avatar-clippedCorner UserBadge-avatar PageHeaderMenu-avatar"
              src="./jacqueline-with-flowers-1954-square.jpg"
            />
          </div>
          <div
            class="Container-flex Container-column"
            style="margin-left: 1em;"
          >
            <div
              class=""
            >
              <h4
                class="MuiTypography-root Typography-invert Typography-headingSmall UserBadge-name PageHeaderMenu-name PageHeaderMenu-truncateText MuiTypography-h4 MuiTypography-displayInline"
              >
                Jacqueline Roque
              </h4>
            </div>
          </div>
        </div>
        <span
          class="DropdownArrow-root PageHeaderMenu-arrow DropdownArrow-medium"
        />
      </div>
    </div>
  </div>
</div>
`;
